radix-engine/src/track/interface.rs

```diff
@@ -23,6 +23,29 @@ pub enum TakeSubstateError {
 
 pub type NodeSubstates = BTreeMap<PartitionNumber, BTreeMap<SubstateKey, IndexedScryptoValue>>;
 
+/// Info about accessing Substate Store Database. Enum contains information which operation
+/// was accessing database including additional associated data.
+#[derive(Default, Debug, Clone)]
+pub enum SubstateStoreDbAccessInfo {
+    AcquireLock,
+    Take,
+    Scan {
+        first_time_read_records_count: usize,
+        read_from_track_count: usize,
+    },
+    TakeMany {
+        first_time_read_records_count: usize,
+        read_from_track_count: usize,
+        update_count: usize,
+    },
+    ScanSorted {
+        first_time_read_records_count: usize,
+        read_from_track_count: usize,
+    },
+    #[default]
+    NoAccess,
```

Comment:
We might want to tie DB access to primitive operations, rather than action.
```rust
pub struct StoreAccessInfo(Vec<StoreAccessInfo>);

// Per substate
pub enum StoreAccess {
    // When store invokes `SubstateDatabase::get_substate()`.
    // size might be zero when the entry does not exist.
    ReadFromDb(usize),
    // When store reads from the `tracked_nodes`.
    ReadFromTrack(usize),
    // When store writes into the `tracked_nodes`.
    // This will eventually be flushed into database.
    Write(usize),
    // When store updates an entry for the second time
    // Need to report both the previous size and new size as partial refund may be required
    // (Reason: only one write will be applied when the transaction finishes, despite substate being updated twice)
    Rewrite(usize, usize),

    // In future, we may want to introduce something like `Clear` to rebate behaviours that 
    // help reduce the state size.
}
```

In case you missed it, the stack looks like:
```
===========
Engine
===========
Track (Store)
===========
Database
===========
````